File: sort/bubble_sort/javascript/bubble-sort.js
const swap = (array, first, second) => {
    
    const temp = array[first];
    
    array[first] = array[second];
    array[second] = temp;
    
    return array;
};

const bubbleSort = (array, nonDecreasing = true) => {
    
    let swapped = true;
    
    while (swapped) {
        
        swapped = false;
        
        for (let i = 0; i < array.length - 1; ++i) {
            
            if (nonDecreasing && (array[i] > array[i + 1]) ||
                !nonDecreasing && (array[i] < array[i + 1])) {

                swap(array, i, i + 1);
                
                swapped = true;
            }
        }
    }
    
    return array;
};

module.exports = bubbleSort;

File: sort/bubble_sort/haskell/bubble_sort.hs
module Bubblesort where
    
    bubblesort::(Ord a)=>[a]->[a]
    bubblesort l = bsort l 0

    swap ::(Ord a)=>[a]->[a]
    swap (x:y:xs)   | x > y = y:swap(x:xs)
                    | otherwise = x:swap(y:xs)
    swap [x] = [x]
    swap [] = []

    bsort::(Ord a)=>[a]->Int->[a]
    bsort l i   | i == (length l) = l
                | otherwise = bsort (swap l) (i + 1)
